zcash_primitives/src/transaction/sighash.rs

```diff
@@ -207,20 +273,53 @@ pub fn signature_hash_data(
             }
             update_u32!(h, tx.lock_time, tmp);
             update_u32!(h, tx.expiry_height.into(), tmp);
-            if sigversion == SigHashVersion::Sapling {
+            if sigversion == SigHashVersion::Sapling || sigversion == SigHashVersion::Future {
                 h.update(&tx.value_balance.to_i64_le_bytes());
             }
             update_u32!(h, hash_type, tmp);
 
-            if let Some((n, script_code, amount)) = transparent_input {
-                let mut data = vec![];
-                tx.vin[n].prevout.write(&mut data).unwrap();
-                script_code.write(&mut data).unwrap();
-                data.extend_from_slice(&amount.to_i64_le_bytes());
-                (&mut data)
-                    .write_u32::<LittleEndian>(tx.vin[n].sequence)
-                    .unwrap();
-                h.update(&data);
+            match signable_input {
+                SignableInput::Transparent {
+                    index,
+                    script_code,
+                    value,
+                } => {
+                    let mut data = if sigversion == SigHashVersion::Future {
+                        ZCASH_TRANSPARENT_SIGNED_INPUT_DOMAIN_SEPARATOR.to_vec()
+                    } else {
+                        vec![]
+                    };
+
+                    tx.vin[index].prevout.write(&mut data).unwrap();
+                    script_code.write(&mut data).unwrap();
+                    data.extend_from_slice(&value.to_i64_le_bytes());
+                    (&mut data)
+                        .write_u32::<LittleEndian>(tx.vin[index].sequence)
+                        .unwrap();
+                    h.update(&data);
+                }
+
+                SignableInput::Tze {
+                    index,
+                    precondition,
+                    value,
+                } if sigversion == SigHashVersion::Future => {
+                    let mut data = ZCASH_TZE_SIGNED_INPUT_DOMAIN_SEPARATOR.to_vec();
+
+                    tx.tze_inputs[index].prevout.write(&mut data).unwrap();
+                    data.write_u32::<LittleEndian>(precondition.extension_id)
+                        .unwrap();
+                    data.write_u32::<LittleEndian>(precondition.mode).unwrap();
+                    data.extend(&precondition.payload);
+                    data.extend_from_slice(&value.to_i64_le_bytes());
+                    h.update(&data);
+                }
+
+                SignableInput::Tze { .. } => {
+                    panic!("A request has been made to sign a TZE input, but the signature hash version is not Future");
```

Comment:
Do we want to call it `Future`? Let's call it something really distinct so that it's easier to replace it with the real name we decide on, in all the places (including this one, which could be accidentally forgotten).

Reply:
Changed to ZFuture.